x-pack/filebeat/input/httpjson/input.go

```diff
@@ -5,596 +5,152 @@
+func Plugin(log *logp.Logger, store cursor.StateStore) input.Plugin {
+	return input.Plugin{
+		Name:       inputName,
+		Stability:  feature.Beta,
+		Deprecated: false,
+		Info:       "HTTP JSON Input",
+		Manager: &cursor.InputManager{
+			Logger:     log.Named(inputName),
+			StateStore: store,
+			Type:       inputName,
+			Configure:  configure,
+		},
 	}
 }
 
-// HttpjsonInput struct has the HttpJsonInput configuration and other userful info.
-type HttpjsonInput struct {
-	config
-	log      *logp.Logger
-	outlet   channel.Outleter // Output of received messages.
-	inputCtx context.Context  // Wraps the Done channel from parent input.Context.
-
-	workerCtx    context.Context    // Worker goroutine context. It's cancelled when the input stops or the worker exits.
-	workerCancel context.CancelFunc // Used to signal that the worker should stop.
-	workerOnce   sync.Once          // Guarantees that the worker goroutine is only started once.
-	workerWg     sync.WaitGroup     // Waits on worker goroutine.
-
-	nextCursorValue string
-}
-
-// RequestInfo struct has the information for generating an HTTP request
-type RequestInfo struct {
-	URL        string
-	ContentMap common.MapStr
-	Headers    common.MapStr
-}
-
-type retryLogger struct {
-	log *logp.Logger
-}
-
-func newRetryLogger() *retryLogger {
-	return &retryLogger{
-		log: logp.NewLogger("httpjson.retryablehttp", zap.AddCallerSkip(1)),
+func configure(cfg *common.Config) ([]cursor.Source, cursor.Input, error) {
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, nil, err
 	}
-}
 
-func (l *retryLogger) Printf(s string, args ...interface{}) {
-	l.log.Debugf(s, args...)
-}
-
-// NewInput creates a new httpjson input
-func NewInput(
-	cfg *common.Config,
-	connector channel.Connector,
-	inputContext input.Context,
-) (input.Input, error) {
-	// Extract and validate the input's configuration.
-	conf := defaultConfig()
-	if err := cfg.Unpack(&conf); err != nil {
-		return nil, err
-	}
-	// Build outlet for events.
-	out, err := connector.Connect(cfg)
+	httpClient, err := newHTTPClient(config)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	// Wrap input.Context's Done channel with a context.Context. This goroutine
-	// stops with the parent closes the Done channel.
-	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
-	go func() {
-		defer cancelInputCtx()
-		select {
-		case <-inputContext.Done:
-		case <-inputCtx.Done():
-		}
-	}()
+	r := &requester{config: config, client: httpClient}
 
-	// If the input ever needs to be made restartable, then context would need
-	// to be recreated with each restart.
-	workerCtx, workerCancel := context.WithCancel(inputCtx)
-
-	in := &HttpjsonInput{
-		config: conf,
-		log: logp.NewLogger("httpjson").With(
-			"url", conf.URL),
-		outlet:       out,
-		inputCtx:     inputCtx,
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-	}
+	in := &httpJSONInput{}
```

Comment:
It looks the requester does too much. It is a 'source', unpacks from the cursor, is a client, and executes the request. Plus the requester gets only partially initialized here. The source is used to allow to generate the key in the key value store, and to pass the input the address where to collect from. No need to confiure an http client yet. Configure should focus on unpacking the config, do not create helper objects that are required for data collection yet.

Following the code your source is the URL only. Maybe you want to consider to add other metadata to the source as well, e.g. parameters, username (in case services return different information based on the login).

Pass the config to the httpJSONInput, not your source. The requester can be created in the "Run" method.

Reply:
I agree, tried to separate pagination, cursor and retry logic out of it, lmk if it still confusing.